src/plugins/discover/public/application/main/components/sidebar/discover_sidebar_responsive.tsx

```diff
@@ -111,38 +122,87 @@ export interface DiscoverSidebarResponsiveProps {
   const [isFlyoutVisible, setIsFlyoutVisible] = useState(false);
-  /**
-   * fieldCounts are used to determine which fields are actually used in the given set of documents
-   */
-  const fieldCounts = useRef<Record<string, number> | null>(null);
-  if (fieldCounts.current === null) {
-    fieldCounts.current = calcFieldCounts(props.documents$.getValue().result!, selectedDataView);
-  }
+  const [sidebarState, dispatchSidebarStateAction] = useReducer(
+    discoverSidebarReducer,
+    selectedDataView,
+    getInitialState
+  );
+  const selectedDataViewRef = useRef<DataView | null | undefined>(selectedDataView);
+  const showFieldList = sidebarState.status !== DiscoverSidebarReducerStatus.INITIAL;
 
-  const [documentState, setDocumentState] = useState(props.documents$.getValue());
   useEffect(() => {
-    const subscription = props.documents$.subscribe((next) => {
-      if (next.fetchStatus !== documentState.fetchStatus) {
-        if (next.result) {
-          fieldCounts.current = calcFieldCounts(next.result, selectedDataView!);
-        }
-        setDocumentState({ ...documentState, ...next });
+    const subscription = props.documents$.subscribe((documentState) => {
+      const isPlainRecordType = documentState.recordRawType === RecordRawType.PLAIN;
+      const dateRange = getResolvedDateRange(data.query.timefilter.timefilter);
+
+      if (documentState?.fetchStatus === FetchStatus.COMPLETE) {
+        dispatchSidebarStateAction({
+          type: DiscoverSidebarReducerActionType.DOCUMENTS_LOADED,
+          payload: {
+            dataView: selectedDataViewRef.current,
+            fieldCounts: calcFieldCounts(documentState.result),
+            isPlainRecord: isPlainRecordType,
+            dateRange,
+          },
+        });
+      } else if (documentState?.fetchStatus === FetchStatus.LOADING) {
+        dispatchSidebarStateAction({
+          type: DiscoverSidebarReducerActionType.DOCUMENTS_LOADING,
+          payload: {
+            dateRange,
+            isPlainRecord: isPlainRecordType,
+          },
+        });
       }
     });
     return () => subscription.unsubscribe();
-  }, [props.documents$, selectedDataView, documentState, setDocumentState]);
+  }, [
+    props.documents$,
+    data.query.timefilter.timefilter,
+    dispatchSidebarStateAction,
+    selectedDataViewRef,
+  ]);
+
+  useEffect(() => {
+    if (selectedDataView !== selectedDataViewRef.current) {
+      dispatchSidebarStateAction({
+        type: DiscoverSidebarReducerActionType.DATA_VIEW_SWITCHED,
+        payload: {
+          dataView: selectedDataView,
+        },
+      });
+      selectedDataViewRef.current = selectedDataView;
+    }
+  }, [selectedDataView, dispatchSidebarStateAction, selectedDataViewRef]);
+
+  const { isProcessing, refetchFieldsExistenceInfo } = useExistingFieldsFetcher({
+    disableAutoFetching: true,
+    dataViews: !isPlainRecord && sidebarState.dataView ? [sidebarState.dataView] : [],
+    query: query!,
+    filters,
```

Comment:
Alright, now I know what's the problem with ignoring pinned filters. they are not part of the `appState`, but of `globalState`, so we need to use filterManager to get all filters like it is done here:

https://github.com/elastic/kibana/blob/1a213bd7cea43caee7ac02f2490c357c27544923/src/plugins/discover/public/application/main/utils/update_search_source.ts#L40